src/analysis/normed_space/normed_group_hom_completion.lean

```diff
@@ -0,0 +1,255 @@
+  refl
+end
+
+lemma normed_group_hom.completion_comp (f : normed_group_hom G H) (g : normed_group_hom H K) :
+  g.completion.comp f.completion = (g.comp f).completion :=
+begin
+  ext x,
+  rw [normed_group_hom.coe_comp, normed_group_hom.completion_def,
+    normed_group_hom.completion_coe_to_fun, normed_group_hom.completion_coe_to_fun,
+    completion.map_comp (normed_group_hom.uniform_continuous _)
+    (normed_group_hom.uniform_continuous _)],
+  refl
+end
+
+lemma normed_group_hom.completion_neg (f : normed_group_hom G H) :
+  (-f).completion = -f.completion :=
+normed_group_hom_completion_hom.map_neg f
+
+lemma normed_group_hom.completion_add (f g : normed_group_hom G H) :
+  (f + g).completion = f.completion + g.completion :=
+normed_group_hom_completion_hom.map_add f g
+
+lemma normed_group_hom.completion_sub (f g : normed_group_hom G H) :
+  (f - g).completion = f.completion - g.completion :=
+normed_group_hom_completion_hom.map_sub f g
+
+@[simp]
+lemma normed_group_hom.zero_completion : (0 : normed_group_hom G H).completion = 0 :=
+normed_group_hom_completion_hom.map_zero
+
+/-- The map from a normed group to its completion, as a normed group hom. -/
+def normed_group.to_compl : normed_group_hom G (completion G) :=
+{ to_fun := coe,
+  map_add' := completion.to_compl.map_add,
+  bound' := ⟨1, by simp [le_refl]⟩ }
+
+open normed_group
+
+lemma normed_group.norm_to_compl (x : G) : ∥to_compl x∥ = ∥x∥ :=
+completion.norm_coe x
+
+lemma normed_group.dense_range_to_compl : dense_range (to_compl : G → completion G) :=
+completion.dense_inducing_coe.dense
+
+@[simp]
+lemma normed_group_hom.completion_to_compl (f : normed_group_hom G H) :
+  f.completion.comp to_compl = to_compl.comp f :=
+begin
+  ext x,
+  change f.completion x = _,
+  simpa
+end
+
+@[simp] lemma normed_group_hom.norm_completion (f : normed_group_hom G H) : ∥f.completion∥ = ∥f∥ :=
+begin
+  apply f.completion.op_norm_eq_of_bounds (norm_nonneg _),
+  { intro x,
+    apply completion.induction_on x,
+    { apply is_closed_le,
+      continuity },
+    { intro g,
+      simp [f.le_op_norm  g] } },
+  { intros N N_nonneg hN,
+    apply f.op_norm_le_bound N_nonneg,
+    intro x,
+    simpa using hN x },
+end
+
+lemma normed_group_hom.ker_le_ker_completion (f : normed_group_hom G H) :
+  (to_compl.comp $ incl f.ker).range ≤ f.completion.ker  :=
+begin
+  intros a h,
+  replace h : ∃ y : f.ker, to_compl (y : G) = a, by simpa using h,
+  rcases h with ⟨⟨g, g_in : g ∈ f.ker⟩, rfl⟩,
+  rw f.mem_ker at g_in,
+  change f.completion (g : completion G) = 0,
+  simp [normed_group_hom.mem_ker, f.completion_coe g, g_in, completion.coe_zero],
+end
+
+lemma normed_group_hom.ker_completion {f : normed_group_hom G H} {C : ℝ}
+  (h : f.surjective_on_with f.range C) :
+  (f.completion.ker : set $ completion G) = closure (to_compl.comp $ incl f.ker).range :=
+begin
+  by_cases Hf : ∀ x, ∥f x∥ = 0, -- This is a bit silly, we simply avoid assuming C ≥ 0
```

Comment:
couldn't you instead have a general lemma saying that, if `f.surjective_on_with f.range C`, then there exists a positive `D` such that `f.surjective_on_with f.range D`? Or even `surjective_on_with.mono`, saying that if it works for `C` then it works for any `D \ge C`?

Reply:
Great idea, thanks!